In [3]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import env
import acquire

from sklearn.impute import SimpleImputer

# Acquire
    Using acuire fuction get_zillow_data

In [4]:
df = acquire.get_zillow_data()
df.head(20)

Reading from csv file...


id  bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  \
0     1         0.0          0.0                           NaN   
1    15         0.0          0.0                           NaN   
2    16         0.0          0.0                           NaN   
3    17         0.0          0.0                           NaN   
4    20         4.0          2.0                        3633.0   
5    27         0.0          0.0                           NaN   
6    31         3.0          4.0                        1620.0   
7    33         3.0          2.0                        2077.0   
8    38         0.0          0.0                           NaN   
9    39         0.0          0.0                           NaN   
10   50         0.0          0.0                           NaN   
11   62         0.0          0.0                        1200.0   
12   71         0.0          0.0                           NaN   
13   76         0.0          0.0                           NaN   
14   97         0.0          0.0                         171.0   
15   99         0.0          0.0                         203.0   
16  105         0.0          0.0                           NaN   
17  106         0.0          0.0                           NaN   
18  110         3.0          1.0                        1244.0   
19  111         3.0          2.0                        1300.0   

    taxvaluedollarcnt  yearbuilt  taxamount    fips  propertylandusetypeid  
0             27516.0        NaN        NaN  6037.0                  261.0  
1                10.0        NaN        NaN  6037.0                  261.0  
2                10.0        NaN        NaN  6037.0                  261.0  
3              2108.0        NaN     174.21  6037.0                  261.0  
4            296425.0     2005.0    6941.39  6037.0                  261.0  
5               124.0        NaN        NaN  6037.0                  261.0  
6            847770.0     2011.0   10244.94  6037.0                  261.0  
7            646760.0     1926.0    7924.68  6037.0                  261.0  
8           6730242.0        NaN   80348.13  6037.0                  261.0  
9             15532.0        NaN     248.89  6037.0                  261.0  
10            11009.0        NaN        NaN  6037.0                  261.0  
11             5328.0     1972.0      91.60  6037.0                  261.0  
12             2171.0        NaN        NaN  6037.0                  261.0  
13              378.0        NaN        NaN  6037.0                  261.0  
14             6920.0     1973.0     255.17  6037.0                  261.0  
15            14166.0     1960.0     163.79  6037.0                  261.0  
16            59927.0        NaN        NaN  6037.0                  261.0  
17            23448.0        NaN        NaN  6037.0                  261.0  
18           169471.0     1950.0    2532.88  6037.0                  261.0  
19           233266.0     1950.0    3110.99  6037.0                  261.0

    Verifiy that all data is property code 261 

In [23]:
df.propertylandusetypeid.value_counts()


261.0    2152863
Name: propertylandusetypeid, dtype: int64

    verify that the is no duplicate data in col: id

In [25]:
df.duplicated(subset=['id']).sum()

0

In [10]:
df.shape

(2152863, 9)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2152863 entries, 0 to 2152862
Data columns (total 9 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   id                            int64  
 1   bedroomcnt                    float64
 2   bathroomcnt                   float64
 3   calculatedfinishedsquarefeet  float64
 4   taxvaluedollarcnt             float64
 5   yearbuilt                     float64
 6   taxamount                     float64
 7   fips                          float64
 8   propertylandusetypeid         float64
dtypes: float64(8), int64(1)
memory usage: 147.8 MB


    Get basic statistics of dataframe

In [15]:
df.describe().T

count          mean            std      min  \
id                            2152863.0  1.490834e+06  860878.470977     1.00   
bedroomcnt                    2152852.0  3.287196e+00       0.954754     0.00   
bathroomcnt                   2152852.0  2.230688e+00       0.999280     0.00   
calculatedfinishedsquarefeet  2144379.0  1.862855e+03    1222.125124     1.00   
taxvaluedollarcnt             2152370.0  4.618962e+05  699676.049600     1.00   
yearbuilt                     2143526.0  1.960950e+03      22.162196  1801.00   
taxamount                     2148421.0  5.634866e+03    8178.910249     1.85   
fips                          2152863.0  6.048377e+03      20.433292  6037.00   
propertylandusetypeid         2152863.0  2.610000e+02       0.000000   261.00   

                                    25%         50%         75%          max  
id                            745117.00  1490613.00  2235859.50   2982282.00  
bedroomcnt                         3.00        3.00        4.00        25.00  
bathroomcnt                        2.00        2.00        3.00        32.00  
calculatedfinishedsquarefeet    1257.00     1623.00     2208.00    952576.00  
taxvaluedollarcnt             188170.25   327671.00   534527.00  98428909.00  
yearbuilt                       1949.00     1958.00     1976.00      2016.00  
taxamount                       2534.98     4108.95     6414.32   1337755.86  
fips                            6037.00     6037.00     6059.00      6111.00  
propertylandusetypeid            261.00      261.00      261.00       261.00

    Get a count of total NaN values by column of DataFrame

In [8]:
df.isna().sum()

id                                 0
bedroomcnt                        11
bathroomcnt                       11
calculatedfinishedsquarefeet    8484
taxvaluedollarcnt                493
yearbuilt                       9337
taxamount                       4442
fips                               0
propertylandusetypeid              0
dtype: int64

In [30]:
df_prep = df.copy(deep=True)

In [31]:
df_prep.dtypes

id                                int64
bedroomcnt                      float64
bathroomcnt                     float64
calculatedfinishedsquarefeet    float64
taxvaluedollarcnt               float64
yearbuilt                       float64
taxamount                       float64
fips                            float64
propertylandusetypeid           float64
dtype: object

In [32]:
df_prep.head(10)

id  bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  \
0   1         0.0          0.0                           NaN   
1  15         0.0          0.0                           NaN   
2  16         0.0          0.0                           NaN   
3  17         0.0          0.0                           NaN   
4  20         4.0          2.0                        3633.0   
5  27         0.0          0.0                           NaN   
6  31         3.0          4.0                        1620.0   
7  33         3.0          2.0                        2077.0   
8  38         0.0          0.0                           NaN   
9  39         0.0          0.0                           NaN   

   taxvaluedollarcnt  yearbuilt  taxamount    fips  propertylandusetypeid  
0            27516.0        NaN        NaN  6037.0                  261.0  
1               10.0        NaN        NaN  6037.0                  261.0  
2               10.0        NaN        NaN  6037.0                  261.0  
3             2108.0        NaN     174.21  6037.0                  261.0  
4           296425.0     2005.0    6941.39  6037.0                  261.0  
5              124.0        NaN        NaN  6037.0                  261.0  
6           847770.0     2011.0   10244.94  6037.0                  261.0  
7           646760.0     1926.0    7924.68  6037.0                  261.0  
8          6730242.0        NaN   80348.13  6037.0                  261.0  
9            15532.0        NaN     248.89  6037.0                  261.0

    convert all zero in columns: ('bedroomcnt', 'bathroomcnt') to NaN

In [33]:
bed_bath_0_columns = ['bedroomcnt', 'bathroomcnt']
df_prep[bed_bath_0_columns] = df_prep[bed_bath_0_columns].replace(0, np.NaN)

    idenify all NaN values in df

In [34]:
df_prep.isnull().sum()

id                                  0
bedroomcnt                      13198
bathroomcnt                     13038
calculatedfinishedsquarefeet     8484
taxvaluedollarcnt                 493
yearbuilt                        9337
taxamount                        4442
fips                                0
propertylandusetypeid               0
dtype: int64

    calculate percentage of NaN values in columns:('bedroomcnt', 'bathroomcnt') to entire dataframe by using len fuction

In [41]:
df_prep[bed_bath_0_columns].isnull().mean() * 100 

bedroomcnt     0.613044
bathroomcnt    0.605612
dtype: float64

In [40]:
df_prep.isnull().mean() * 100

id                              0.000000
bedroomcnt                      0.613044
bathroomcnt                     0.605612
calculatedfinishedsquarefeet    0.394080
taxvaluedollarcnt               0.022900
yearbuilt                       0.433702
taxamount                       0.206330
fips                            0.000000
propertylandusetypeid           0.000000
dtype: float64

    eventhough about 60% of col: 'bedroomcnt', 'bathroomcnt' is NaN values comparied to whole dataframe. Droping the rows with NaN values (13,198). Which will still leave about 40% to make a prediction on property value with a total of 2,152,863 unique home rows 

In [14]:
df_prep = df_prep.dropna(subset=['bedroomcnt', 'bathroomcnt'], axis=0)
df_prep.reset_index(drop=True)

id  bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  \
0             20         4.0          2.0                        3633.0   
1             31         3.0          4.0                        1620.0   
2             33         3.0          2.0                        2077.0   
3            110         3.0          1.0                        1244.0   
4            111         3.0          2.0                        1300.0   
...          ...         ...          ...                           ...   
2138721  2982265         4.0          4.0                        4375.0   
2138722  2982268         4.0          3.0                        2262.0   
2138723  2982270         4.0          4.5                        3127.0   
2138724  2982275         3.0          2.5                        1974.0   
2138725  2982282         4.0          4.0                        2110.0   

         taxvaluedollarcnt  yearbuilt  taxamount    fips  \
0                 296425.0     2005.0    6941.39  6037.0   
1                 847770.0     2011.0   10244.94  6037.0   
2                 646760.0     1926.0    7924.68  6037.0   
3                 169471.0     1950.0    2532.88  6037.0   
4                 233266.0     1950.0    3110.99  6037.0   
...                    ...        ...        ...     ...   
2138721           422400.0     2015.0   13877.56  6037.0   
2138722           960756.0     2015.0   13494.52  6059.0   
2138723           536061.0     2014.0    6244.16  6059.0   
2138724           424353.0     2015.0    5302.70  6059.0   
2138725           554009.0     2014.0    6761.20  6037.0   

         propertylandusetypeid  
0                        261.0  
1                        261.0  
2                        261.0  
3                        261.0  
4                        261.0  
...                        ...  
2138721                  261.0  
2138722                  261.0  
2138723                  261.0  
2138724                  261.0  
2138725                  261.0  

[2138726 rows x 9 columns]

In [27]:
df_prep.shape

(2138726, 9)

    confirm that NaN values in cols 'bedroomcnt', 'bathroomcnt' have been dropped and see where other NaN values to be simple imputation(mean, median, mode) or Iterative Imputation

In [28]:
df_prep.isnull().sum()

id                                 0
bedroomcnt                         0
bathroomcnt                        0
calculatedfinishedsquarefeet     211
taxvaluedollarcnt                107
yearbuilt                       1250
taxamount                       2490
fips                               0
propertylandusetypeid              0
dtype: int64

In [26]:
df_prep.dtypes

id                                int64
bedroomcnt                      float64
bathroomcnt                     float64
calculatedfinishedsquarefeet    float64
taxvaluedollarcnt               float64
yearbuilt                       float64
taxamount                       float64
fips                            float64
propertylandusetypeid           float64
dtype: object

In [19]:
df_impute=df_prep.copy(deep=True)

In [20]:
cols_to_impute = ['calculatedfinishedsquarefeet', 'taxvaluedollarcnt', 'yearbuilt', 'taxamount']

    Now to check the columns that have NaN values what is the skew to determine what to use to impute (mean, median, mode)

In [21]:
df_impute[cols_to_impute].skew()

calculatedfinishedsquarefeet    257.762625
taxvaluedollarcnt                18.735569
yearbuilt                         0.069680
taxamount                        19.393004
dtype: float64

In [ ]:
edf = SimpleImputer

In [ ]:
edf.kurd 